### DISCUSSION TITLE
Payload Form Builder Custom Fields

### DISCUSSION THREAD

**Discussion by mabedd** [payloadcms/payload#7161] [Q&A]
Hi,
I started using form builder plugin and it seems very nice. 

But my question is there any ideas for adding custom fields like for file upload radio or anything ?

**Comment by dkirchhof** [comment]
While the docs don't mention anything about custom fields, here is a small example:

```ts
formBuilderPlugin({
  fields: {
    // ...
    // this is of type Block
    date: {
      slug: 'date',
      labels: {
        singular: 'Date',
        plural: 'Dates',
      },
      fields: [
        {
          name: 'date',
          label: 'Date',
          type: 'date',
        },
      ],
    },
  },
}),
```

The default fields (like text, email, etc.) have more "inner fields". You can copy them from `packages/plugin-form-builder/src/collections/Forms/fields.ts`

**Comment by GodGardensGuns** [comment]
Thanks @dkirchhof, that worked for me.

What's strange is, the type definition of a custom field for this plugin is `Partial<Field>`, but passing a `Field` object throws an error:

```js
TypeError: Cannot read properties of undefined (reading 'concat')
```

The custom field has to be a `Block` per the documentation [here](https://payloadcms.com/docs/plugins/form-builder#field-overrides) (and per your advice @dkirchhof). To avoid TypeScript throwing errors, I'm having to add `@ts-expect-error` to my code.

**Reply by mikecebul** [reply]
In the admin panel the form "fields" are actually blocks with several fields to select all the options for each individual "field".
